packages/noya-state/src/reducers/history.ts

```diff
@@ -0,0 +1,99 @@
+import produce from 'immer';
+import { ApplicationState, Action, reducer } from './application';
+import { createInitialInteractionState } from './interaction';
+
+export type HistoryEntry = {
+  actionType: Action[0];
+  timestamp: number;
+  state: ApplicationState;
+};
+
+export type HistoryState = {
+  past: HistoryEntry[];
+  present: ApplicationState;
+  future: HistoryEntry[];
+};
+
+export type HistoryAction = [type: 'undo'] | [type: 'redo'] | Action;
+
+const FILE_CHANGED_TIMEOUT = 300;
+
+export function historyReducer(state: HistoryState, action: HistoryAction) {
+  const currentPresent = state.present;
+  const [actionType] = action;
+  switch (actionType) {
+    case 'undo':
+      return produce(state, (state) => {
+        if (state.past.length > 0) {
+          const newPresent = state.past.pop() as HistoryEntry;
```

Comment:
We should figure some way of avoiding the `as` cast here, even if it means doing some work before we call `produce`

Reply:
Probably we want to do `if (state.past.length === 0) { return state }` before `produce` regardless